File: app/src/main/java/org/schulcloud/mobile/util/crypt/JWTUtil.java
package org.schulcloud.mobile.util.crypt;

import android.util.Base64;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.UnsupportedEncodingException;


public class JWTUtil {
    private static final String KEY_USER_ID = "userId";

    public static String decodeToCurrentUser(String JWTEncoded) {
        try {
            String[] split = JWTEncoded.split("\\.");

            // get body of jwt
            JsonParser jsonParser = new JsonParser();
            String bodyJson = getJson(split[1]);
            JsonObject jsonObject = jsonParser.parse(bodyJson).getAsJsonObject();

            return jsonObject.get(KEY_USER_ID).getAsString();
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
            return null;
        }
    }

    private static String getJson(String strEncoded) throws UnsupportedEncodingException {
        byte[] decodedBytes = Base64.decode(strEncoded, Base64.URL_SAFE);
        return new String(decodedBytes, "UTF-8");
    }
}

File: app/src/main/java/org/schulcloud/mobile/utils/JWTUtil.kt
@file:Suppress("TooManyFunctions")

package org.schulcloud.mobile.utils

import android.util.Base64
import com.google.gson.JsonParser
import java.io.UnsupportedEncodingException
import java.nio.charset.Charset

object JwtUtil {
    private const val KEY_USER_ID = "userId"

    fun decodeToCurrentUser(jwtEncoded: String): String? {
        @Throws(UnsupportedEncodingException::class)
        fun String.decodeBase64(): String {
            val decodedBytes = Base64.decode(this, Base64.URL_SAFE)
            return String(decodedBytes, Charset.forName("UTF-8"))
        }

        return try {
            val bodyRaw = jwtEncoded.split("\\.".toRegex())
                    .dropLastWhile { it.isEmpty() }[1]
                    .decodeBase64()
            JsonParser().parse(bodyRaw).asJsonObject
                    .get(KEY_USER_ID).asString
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
            null
        }
    }
}
